src/dbSta/test/hierwrite.tcl

```diff
@@ -0,0 +1,7 @@
+source "helpers.tcl"
+read_lef Nangate45/Nangate45.lef
+read_liberty Nangate45/Nangate45_typ.lib
+read_db hierwrite.odb -hier
```

Comment:
Is there a way to use a def file here instead of reading an odb file? It would be better for stability, as future odb schema changes might break this test.

Reply:
You are aware that def is inherently flat, right ? And that the def reader does not know anything about hierarchy ?
Are you saying that the verilog written should somehow be dependent on the way the database is built and that is acceptable ?

I agree that the open road verilog writer (which I have found a few bugs in and had fixed by Cherry, for example the accumulation of child module instances was wrong and was fixed a couple of weeks ago), might have further issues and possibly not sort its objects to preserve stability. But that should be the subject of another pull request. Likewise Peter's request for a quality verilog writer should be the subject of another pull request.

Perhaps one way forward here is to simply remove the test which I think has unnecessarily distracted you and Peter. If that is done will the pull request be ok ?

I would point out to both of you that this pull request was simply issued to add one missing api call to the dbNetwork so that entire modules were not missed (!) in the OpenRoad verilog writer so that the module swapping project could proceed.

Reply:
> Are you saying that the verilog written should somehow be dependent on the way the database is built and that is acceptable ?

Just to clarify that, no. I didn't say that. If the odb schema is updated, this odb might be outdated and not properly loaded. That's why almost every unit test we have uses LEF/DEF.

Reply:
I think this does point out an area where we need to think about how to best address integration tests with hierarchy. LEF/DEF doesn't support it, and ODB is unstable. We need to think of something that is both stable and has hierarchy. Perhaps we could have a CI task that checks all the ODB files are generated with the latest version and have an update script that those who modify ODB will have to run to update them to the latest version

Reply:
Thanks.

Perhaps a fair compromise on the hierarchical tests is as soon as  possible clean up the **hierarchical** verilog writer so it is assured to be invariant of module/port order and db construction (no dependence on ids).  I hasten to add the verilog writer as is, has not changed for flat designs.  I suspect this might address Peter's requirements too (a separate pull request for pretty verilog writing could be made)

So the hierarchical tests would start with odb, sdc, liberty and then write out invariant verilog (by invariant I mean the module order, instance order, net -- properly named !-  order, port orders all sorted by name, with the root at the top). Now sure that won't test much other than the logical structure, but it is a start and for sure if someone changes the logical structure the golden output verilog  would need to regoldened after formal verification.

Reply:
ps one small thing there is a C++ unit test case which builds a hierarchical test case and writes out both the flat and hierarchical view of the odb as a string and then does a string diff to validate that the odb has not changed. This might be of use (see function odb/test/cpp/TestModule.cpp: DbStrDebugHierarchy). So there is a little coverage on the odb schema/hierarchy stuff, but I agree not very much. That function might have some utility for validating odbs not getting screwed up by changes (eg read_db design.odb -hier, then write out using some version of DbStrDebugHierarchy to make sure nets, instance etc unchanged).